src/peft/tuners/oft/layer.py

```diff
@@ -0,0 +1,378 @@
+
+    def get_delta_weight(self, adapter_name: str) -> torch.Tensor:
+        scaling = self.scaling[adapter_name]
+        rank = self.r[adapter_name]
+        coft = self.coft[adapter_name]
+        eps = self.eps[adapter_name]
+        opt_r = self.oft_r[adapter_name]
+
+        if coft:
+            with torch.no_grad():
+                opt_r.copy_(self._project_batch(opt_r, eps=eps))
+
+        orth_rotate = self._cayley_batch(opt_r * scaling)
+        weight = self._block_diagonal(orth_rotate, rank)
+
+        return weight
+
+    # Copied from https://github.com/Zeju1997/oft/blob/84cebb965df69781e3d9c3c875f5980b421eaf24/oft-control/oft.py#L144
+    def _cayley_batch(self, data: torch.Tensor) -> torch.Tensor:
+        b, r, c = data.shape
+        # Ensure the input matrix is skew-symmetric
+        skew = 0.5 * (data - data.transpose(1, 2))
+        I = torch.eye(r, device=data.device).unsqueeze(0).expand(b, r, c)
+
+        # Perform the Cayley parametrization
+        Q = torch.bmm(I - skew, torch.inverse(I + skew))
+
+        return Q
+
+    # Copied from https://github.com/Zeju1997/oft/blob/84cebb965df69781e3d9c3c875f5980b421eaf24/oft-control/oft.py#L155
+    def _block_diagonal(self, oft_r: torch.Tensor, rank: int) -> torch.Tensor:
+        if oft_r.shape[0] == 1:
+            # block share
+            blocks = [oft_r[0, ...] for i in range(rank)]
+        else:
+            blocks = [oft_r[i, ...] for i in range(rank)]
+
+        # Use torch.block_diag to create the block diagonal matrix
+        A = torch.block_diag(*blocks)
+
+        return A
+
+    # Copied from https://github.com/Zeju1997/oft/blob/84cebb965df69781e3d9c3c875f5980b421eaf24/oft-control/oft.py#L52
+    def _project_batch(self, oft_r, eps=1e-5):
+        # scaling factor for each of the smaller block matrix
+        eps = eps * 1 / torch.sqrt(torch.tensor(oft_r.shape[0]))
+        I = (
+            torch.zeros((oft_r.size(1), oft_r.size(1)), device=oft_r.device, dtype=oft_r.dtype)
+            .unsqueeze(0)
+            .expand_as(oft_r)
+        )
+        diff = oft_r - I
+        norm_diff = torch.norm(oft_r - I, dim=(1, 2), keepdim=True)
+        mask = (norm_diff <= eps).bool()
+        out = torch.where(mask, oft_r, I + eps * (diff / norm_diff))
+        return out
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            result = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            result = self.base_layer(x, *args, **kwargs)
+        else:
+            result = self.base_layer(x, *args, **kwargs)
+            if len(result.shape) == 4:
+                result = result.permute(0, 2, 3, 1)
+
+            base_layer = self.get_base_layer()
+            base_bias = base_layer.bias
+            if base_bias is not None:
+                result = result - base_bias.data
```

Comment:
Not super happy that this is here. I assume it is necessary for correctness? Could you please explain and maybe add a comment?

Reply:
@BenjaminBossan In my quick check, there are some differences.

```
import torch


def cayley_batch(data):
        b, r, c = data.shape
        # Ensure the input matrix is skew-symmetric
        skew = 0.5 * (data - data.transpose(1, 2))
        # I = torch.eye(r, device=data.device).unsqueeze(0).repeat(b, 1, 1)
        I = torch.eye(r, device=data.device).unsqueeze(0).expand(b, r, c)

        # Perform the Cayley parametrization
        Q = torch.bmm(I - skew, torch.inverse(I + skew))

        return Q

def block_diagonal(R):
        if len(R.shape) == 2:
            # Create a list of R repeated block_count times
            blocks = [R] * r
        else:
            # Create a list of R slices along the third dimension
            blocks = [R[i, ...] for i in range(r)]

        # Use torch.block_diag to create the block diagonal matrix
        A = torch.block_diag(*blocks)

        return A

in_features = 128
r = 4
eps = 6e-5
R_shape = [r, in_features // r, in_features // r]
R = torch.zeros(R_shape[1], R_shape[1])
R = torch.stack([R] * r)
eps = eps * R_shape[1] * R_shape[1]

# generate random R
orth_rotate = cayley_batch(torch.rand(*R.shape) * 0.001)
block_diagonal_matrix = block_diagonal(orth_rotate)

from torch import nn
attn = nn.Linear(32, 128, bias=True)
fix_filt = attn.weight.data
base_bias = attn.bias.data if attn.bias is not None else None
filt = torch.mm(torch.transpose(fix_filt, 0, 1), block_diagonal_matrix)
filt = torch.transpose(filt, 0, 1)

x = torch.rand(1, 32)

out = x
out = nn.functional.linear(input=out, weight=filt, bias=base_bias)

out2 = attn(x)
out2 = torch.matmul(out2, block_diagonal_matrix)

out3 = attn(x) - base_bias
out3 = torch.matmul(out3, block_diagonal_matrix) + base_bias
```

out == out3, out2 != out3

Reply:
> out2 = attn(x)
> out2 = torch.matmul(out2, block_diagonal_matrix)
> out3 = attn(x) - base_bias
> out3 = torch.matmul(out3, block_diagonal_matrix) + base_bias

Yes, a difference is expected when in one case, we first add the bias, then multiply, in the other we first multiply, then add the bias. I understand why we would first want to accumulate all the multiplications before applying the bias, it's just unfortunate that it means we have to manually remove and add the bias, which I think could be error prone. I have no solution really, except to add a comment why this is done.